R/hardcode.R

```diff
@@ -13,11 +13,26 @@ sdtm_hardcode <- function(raw_dat,
 
   tgt_val <- ct_map(tgt_val, ct = ct, cl = cl)
 
-  raw_dat |>
-    dplyr::right_join(y = tgt_dat, by = by) |>
-    dplyr::mutate("{tgt_var}" := recode(x = !!rlang::sym(raw_var), to = tgt_val)) |>
-    dplyr::select(-rlang::sym(raw_var)) |>
-    dplyr::relocate(tgt_var, .after = dplyr::last_col())
+  # When target dataset and  by variables are provided,
+  # the tar_var is added to the input target dataset.
+  if((!is.null(by))){
+    tgt_dat_out <- raw_dat |>
+      #we need to keep only the required variables in the input raw dataset
+      dplyr::select(dplyr::all_of(by), rlang::sym(raw_var)) |>
+      dplyr::right_join(y = tgt_dat, by = by) |>
+      dplyr::mutate("{tgt_var}" := recode(x = !!rlang::sym(raw_var), to = tgt_val)) |>
+      dplyr::select(-rlang::sym(raw_var)) |>
+      dplyr::relocate(tgt_var, .after = dplyr::last_col())
+  } else {
+    # When target dataset and  by variables are NOT provided,
+    # the tgt_dat_out is created with tar_var & oak_id_vars.
+    tgt_dat_out <- raw_dat |>
+      dplyr::select(oak_id, raw_source, patient_number,rlang::sym(raw_var)) |>
```

Comment:
`oak_id`, `raw_source`, `patient_number` do not exist in the context of this function, did you mean `dplyr::select(by, rlang::sym(raw_var))`?

Reply:
Yes, you are correct. We need to create them as oak_id_vars.  We need to create something like this and refer to it in all functions https://github.com/pharmaverse/roak_pilot/blob/main/R/zzz.R#L17

I will explain the use of this in the meeting tomorrow.

Reply:
I understand that there might be special columns that work as keys and that might be expected, but isn't it better to pass them as an argument instead of having them as a global variable? Having functions depending on global state is generally a bad idea.

And also, if sdtm.oak is meant to be EDC agnostic then the RAVE_SOURCE (is this about Medidata’s Rave EDC?) ought to be changed to something else, right?

Reply:
RAVE_SOURCE is renamed as raw_source in the open source {sdtm.oak}
I am okay with passing them as parameters. We have to pass this default parameter again and again. Whichever works better, we can do it.

Reply:
Well, if it is a default parameter then it won't be actually passing again and again. It will be in the function definition but the user and developer won't be explicitly writing down those values.

Reply:
Decided to create a function to deal with oak_id, raw_source, patient_number